### PR TITLE
feat: add istio course

### PR DESCRIPTION
**Pull request by ArchiFleKs**

### CODE REVIEW

**Comment by ArchiFleKs** [comment]
This is only the course I'll add TPs in separate PR

**Comment by ArchiFleKs** [comment]
Actually, TPs are in this PR. This would need to be merged tomorrow.

File: cours/containers/istio/istio-introduction.fr.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+- Map des communications entre micro services
+- Tracing applicatif
+- Monitoring et logging avec possibilité de réaliser des dashboard personnalisés
+- Détection d'erreurs
+
+### Plateforme
+
+Istio fonctionne nativement avec Kubernetes mais peut s'intégrer à d'autres systèmes comme :
+
+- Service enregistrés avec Consul
```
Comment (rguichard) [inline cours/containers/istio/istio-introduction.fr.md:49]: "services enregistrés" ?

File: cours/containers/istio/istio-introduction.fr.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
+# Istio: Introduction
+
+### Présentation
+
+- Fourni une implémentation du concept de service mesh
```
Comment (rguichard) [inline cours/containers/istio/istio-introduction.fr.md:5]: fournit

File: cours/containers/istio/istio-introduction.fr.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+![](images/istio/arch.svg){height="400px"}
+
+### Management du trafic
+
+- Configuration de règles de routage applicatives :
+    - Timeout
+    - Retry
+    - Circuit Breaker
+- Simplifie la mise en place de mise à jour automatiques :
```
Comment (rguichard) [inline cours/containers/istio/istio-introduction.fr.md:27]: mises

File: cours/containers/istio/istio-introduction.fr.md [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+### Présentation
+
+- Fourni une implémentation du concept de service mesh
+- Open Sourcé en 2017 par Google, IBM et Lyft
+
+![](images/istio/logo.png){height="200px"}
+
+### Fonctionnalités
+
+- Load balancing  TCP, GRPC, HTTP et websocket
```
Comment (rguichard) [inline cours/containers/istio/istio-introduction.fr.md:12]: gRPC

File: cours/containers/istio/istio-architecture.fr.md [hunk 5]
```diff
@@ -0,0 +1,47 @@
+# Istio : Architecture
+
+### Composants
+
+- Plan de donnée : composé de proxy [Envoy](https://www.envoyproxy.io/)
```
Comment (rguichard) [inline cours/containers/istio/istio-architecture.fr.md:5]: - données

File: cours/containers/istio/istio-architecture.fr.md [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+- Plan de contrôle: `Istiod` qui pilote la configuration du plan de donnée
```
Comment (rguichard) [inline cours/containers/istio/istio-architecture.fr.md:6]: space before `:`

File: cours/containers/istio/istio-architecture.fr.md [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+### Topologies
+
+Istio supporte de multiples méthodes de déploiement :
+
+- un ou plusieurs cluster
```
Comment (rguichard) [inline cours/containers/istio/istio-architecture.fr.md:12]: clusters

File: cours/containers/istio/istio-architecture.fr.md [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+### Topologies
+
+Single cluster : la configuration la plus simple et communément utilisée
+
+![](images/istio/single-cluster.svg){height="400px"}
+
+### Topologies
+
+Multi cluster:
```
Comment (rguichard) [inline cours/containers/istio/istio-architecture.fr.md:27]: space before `:`

File: cours/containers/istio/istio-architecture.fr.md [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+![](images/istio/multi-cluster.svg){height="400px"}
+
+### Sidecar
+
+Un sidecar est un conteneur fonctionnant à coté d'un autre conteneur au sein
+d'un même pods. Dans un mesh Istio, chaque pods dispose d'un conteneur Envoy
```
Comment (rguichard) [inline cours/containers/istio/istio-architecture.fr.md:34]: pod

File: cours/containers/istio/istio-architecture.fr.md [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+code de l'application n'a pas besoin d'être modifié et que l'intégration
+devient transparente pour l'application.
+
+### Sidecar
+
+Il existe plusieurs façons d'injecter ce sidecar dans les pods:
+
+- manuellement via `istioctl`
+- automatiquement via Kubernetes et un `MutatingAdmissionWebhook`
+- automatiquement via webhook + CNI: recommandée
```
Comment (rguichard) [inline cours/containers/istio/istio-architecture.fr.md:45]: space before `:`

File: cours/containers/istio/istio-observability.fr.md [hunk 11]
```diff
@@ -0,0 +1,29 @@
+# Istio: Observabilité
+
+### Observabilité
+
+Istio fournit par défaut un service de télémétrie pour toutes les communication au sein du mesh:
```
Comment (rguichard) [inline cours/containers/istio/istio-observability.fr.md:5]: space before `:`

File: cours/containers/istio/istio-observability.fr.md [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+
+- Métriques applicative: dashboard et compatibilité [prometheus](https://prometheus.io/)
```
Comment (rguichard) [inline cours/containers/istio/istio-observability.fr.md:7]: applicatives :

File: cours/containers/istio/istio-observability.fr.md [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+- Traçage applicatif : integration avec les solutions les plus utilisées:
```
Comment (rguichard) [inline cours/containers/istio/istio-observability.fr.md:8]: space before `:`

File: cours/containers/istio/istio-security.fr.md [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+Istio fournit des composants pour :
+
+- Se défendre contre les attaques *man in the middle* : chiffrement
+- Mettre en place des politiques d'accès par authentification et autorisation
+- Disposer d'un audit des flux au sein du mesh
+
+### Sécurité
+
+Ces fonctionnalités sont principalement mises en place grâce au
+[mTLS](https://en.wikipedia.org/wiki/Mutual_authentication): mutual TLS entre
```
Comment (rguichard) [inline cours/containers/istio/istio-security.fr.md:14]: space before `:`

File: cours/containers/istio/istio-security.fr.md [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+### Sécurité
+
+![](images/istio/id-prov.svg){height="400px"}
+
+### Sécurité : authentification
+
+Istio fourni 2 méthodes d'authentification :
+
+- Service to Service: via mTLS
+- End User: via des fournisseur d'identités compatibles avec [OpenID connect](https://openid.net/connect/):
```
Comment (rguichard) [inline cours/containers/istio/istio-security.fr.md:37]: fournisseurs

File: cours/containers/istio/istio-security.fr.md [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+    - Keycloak
+    - Auth0
+    - Firebase
+    - Google Auth
+    - ORY Hydra
+
+### Sécurité : mTLS
+
+Le chiffrement du trafic ainsi que l'authentification est réalisé via mTLS. Il
+existe 3 modes:
```
Comment (rguichard) [inline cours/containers/istio/istio-security.fr.md:47]: space before `:`

File: cours/containers/istio/istio-security.fr.md [hunk 17]
```diff
@@ -0,0 +1,138 @@
...
+  selector:
+    matchLabels:
+      app: reviews
+  mtls:
+    mode: STRICT
+```
+
+### Sécurité : Authentication Policy
+
+S'appliquent:
```
Comment (rguichard) [inline cours/containers/istio/istio-security.fr.md:79]: space before `:`

File: cours/containers/istio/istio-security.fr.md [hunk 18]
```diff
@@ -0,0 +1,138 @@
...
+
+- Au niveau du cluster/mesh global: non namespacées
+- Au sein d'un namespace: namespacées
+- Au sein d'un namespace: namespacées + selector :
+
```
Comment (rguichard) [inline cours/containers/istio/istio-security.fr.md:84]: space before `:` for the three previous lines

File: cours/containers/istio/istio-security.fr.md [hunk 19]
```diff
@@ -0,0 +1,138 @@
...
+### Sécurité : Autorisation
+
+En plus de l'authentification, Istio fournit une couche d'autorisation qui permet de définir des règles de sécurité applicatives
+
+![](images/istio/authn.svg){height="400px"}
+
+### Sécurité : Authorization policies
+
+- Fonctionnent avec les protocoles HTTP, HTTPS, HTTP2, GRPC et TCP
```
Comment (rguichard) [inline cours/containers/istio/istio-security.fr.md:99]: gRPC

File: cours/containers/istio/istio-traffic.fr.md [hunk 20]
```diff
@@ -0,0 +1,310 @@
...
+        subset: v2
+  - route:
+    - destination:
+        host: reviews
+        subset: v3
+```
+
+### Virtual Services
+
+Routage entre deux versions d'un même service avec pourcentage:
```
Comment (rguichard) [inline cours/containers/istio/istio-traffic.fr.md:60]: space before `:`

File: cours/containers/istio/istio-traffic.fr.md [hunk 21]
```diff
@@ -0,0 +1,310 @@
...
+    route:
+    - destination:
+        host: ratings
+```
+
+### Destination Rules
+
+- Contrôlent les opérations réalisées sur le trafic pour une destination donnée
+- Groupement des services en *subset* pour utilisation avec les *VirtualServices*
+- Techniques de répartition de charge:
```
Comment (rguichard) [inline cours/containers/istio/istio-traffic.fr.md:109]: space before `:`

File: cours/containers/istio/istio-traffic.fr.md [hunk 22]
```diff
@@ -0,0 +1,310 @@
...
+  location: MESH_EXTERNAL
+  resolution: DNS
+```
+
+### Sidecars
+
+Par défaut les sidecar disposent de la connaissance et de la possibilité de
+joindre tous les services du mesh. Il est possible de limiter les services
+joignables par certains sidecars, par exemple dans le cas de larges clusters où
+la consommation CPU/RAM des sidecar peut devenir non négligeable.
```
Comment (rguichard) [inline cours/containers/istio/istio-traffic.fr.md:200]: sidecars

File: cours/containers/istio/istio-traffic.fr.md [hunk 23]
```diff
@@ -0,0 +1,310 @@
...
+        host: ratings
+        subset: v1
+    retries:
+      attempts: 3
+      perTryTimeout: 2s
+```
+
+### Résilience et tests
+
+- Circuit Breakers : permettent de limiter le nombre de connections à un service en fonction de paramètres afin d'éviter une surcharge
```
Comment (rguichard) [inline cours/containers/istio/istio-traffic.fr.md:261]: connexions

File: cours/containers/istio/istio-traffic.fr.md [hunk 24]
```diff
@@ -0,0 +1,310 @@
...
+      version: v1
+    trafficPolicy:
+      connectionPool:
+        tcp:
+          maxConnections: 100
+```
+
+### Résilience et tests
+
+Fault Injection:
```
Comment (rguichard) [inline cours/containers/istio/istio-traffic.fr.md:282]: space before `:`

**Review by rguichard — COMMENTED** [review]
space before `:`
space before `:`
space before `:`
space before `:`
space before `:`
space before `:`

File: cours/containers/istio/istio-introduction.fr.md (hunk 3 above)
Comment (ArchiFleKs) [reply cours/containers/istio/istio-introduction.fr.md:27]: la mise ?

Comment (ArchiFleKs) [reply cours/containers/istio/istio-introduction.fr.md:27]: une mise en place

Comment (ArchiFleKs) [reply cours/containers/istio/istio-introduction.fr.md:27]: les mises a jour ok ^^

**Review by rguichard — CHANGES_REQUESTED** [review]
A dedicated Vagrant env should be created for istio.

**Review by rguichard — APPROVED** [review]
lgtm

**Pull request merged**